### PR TITLE
MAG catalogue search improvements; db + development experience improvements

### PR DESCRIPTION
**Pull request by SandyRogers**
This PR:
- Ensures a fresh database can be created by code, that matches the current reality:
  - squashes `emgapi` migrations 1 to 43.
  - introduces various fixes and `RunPython` migration steps to cover for some historic "by-hand" database changes, and pre-Django database states, in the new initial migration.
  - adds or fixes several new management commands to populate database tables for Biome, Variable Names, Analysis Status etc. These table states are long-lived – initial data is in JSON fixtures.
- Adds the ability to create a minimal fake SQLite ENA database, just enough for some testing fixtures. This is to improve local development experience and CI, especially:
  - the `import_analysis` can now be run without connecting to internal ENA databases
  - when we switch entirely to ENA API instead of ENA db access, it will be easier to dev
- Makes EMG API able to work with SQLite
  - Couple of fixes for MySQL specific things (especially creating indices in the squashed migration)
  - New dockerfile for building EMG API to work with SQLite
- Adds features to MAGs (genomes):
  - Enables (or fixes) cross-catalogue MAG searching with COBS and Sourmash
  - adds some extra fields to MAG catalogues to display more structured data about the catalogues (like genomes annotation pipeline version)
 
This goes together with [changes on `mgnify-web`](https://github.com/EBI-Metagenomics/mgnify-web), including a Taskfile and new docker-compose setup.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -11,12 +11,13 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.6, 3.7.1, 3.8.6]
+        python-version: [3.6, 3.8.6]
+        # TODO: Temporarily removed 3.7.1, waiting for https://github.com/celery/celery/issues/7783
```
Comment (mberacochea) [inline .github/workflows/test.yml:15]: We should review our envs and bump everything to >= 3.8

File: emgapi/backends.py [hunk 2]
```diff
@@ -38,7 +39,10 @@ def authenticate(self, request, username=None, password=None):
         }
         req = requests.post(ena_auth_url, json=data)
         logging.info(f'Response from ENA auth: {req}')
-        resp = req.json()
+        try:
+            resp = req.json()
+        except JSONDecodeError:
```
Comment (mberacochea) [inline emgapi/backends.py:44]: Shouldn't we log this error?

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 3]
```diff
@@ -0,0 +1,2489 @@
...
+
+"""
+This squashed migration creates the EMG database as it was at migration 0043 (2022-08-11), 
+whilst populating certain tables to the state they were at that time.
+Those tables are things like Download Description Labels, whose data is (should be) updated through
+migrations.RunPython steps because some code (e.g. import scripts) relies on hardcoded matches to these data.
+
+Some steps are RunSQL and assume SQL dialect that only works on MySQL. 
+These steps have been changed to a helper class (`RunSqlIfMySQL`) so that they can be run on other engines, e.g. SQLite.
+This is relevant to the creation of Full Text Indexes.
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:19]: sqlite full text plugin -> https://www.sqlite.org/fts5.html
note saying we should use it... sqlite has it all hehe

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 4]
```diff
@@ -0,0 +1,2489 @@
...
+    """
+    genecluster_types = [
+        ['acyl_amino_acids', 'N-acyl amino acid cluster'],
+        ['amglyccycl', 'Aminoglycoside/aminocyclitol cluster'],
+        ['aminocoumarin', 'Aminocoumarin cluster'],
+        ['arylpolyene', 'Aryl polyene cluster'],
+        ['bacteriocin', 'Bacteriocin or other unspecified ribosomally ' +
+                        'synthesised and post-translationally modified peptide product (RiPP) cluster'],
+        ['betalactone', 'beta-lactone containing protease inhibitor'],
+        ['blactam', 'Î²-lactam cluster'],
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:324]: Is this char OK?

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py (hunk 3 above)
Comment (SandyRogers) [reply emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:19]: @mberacochea Yeh for sure :) My thinking here was that for sqlite test purposes we don't really need the FTIs (tiny data volume) and that for production deployments we would probably need engine-specific migrations for mysql/postgres etc...

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 5]
```diff
@@ -0,0 +1,2489 @@
...
+                ('study', models.ForeignKey(db_column='STUDY_ID', on_delete=django.db.models.deletion.CASCADE, to='emgapi.study')),
+                ('sample', models.ForeignKey(db_column='SAMPLE_ID', on_delete=django.db.models.deletion.CASCADE, to='emgapi.sample')),
+            ],
+            options={
+                'db_table': 'STUDY_SAMPLE',
+                'unique_together': {('study', 'sample')},
+            },
+        ),
+        # migrations.RunSQL(
+        #     sql='INSERT INTO STUDY_SAMPLE (STUDY_ID, SAMPLE_ID) SELECT STUDY_ID, SAMPLE_ID from SAMPLE where SAMPLE_ID>0',
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:923]: Dev derbis

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 6]
```diff
@@ -0,0 +1,2489 @@
...
+            model_name='analysisjob',
+            name='sample',
+            field=models.ForeignKey(db_column='SAMPLE_ID', on_delete=django.db.models.deletion.CASCADE, related_name='analysis', to='emgapi.sample'),
+        ),
+        migrations.AddField(
+            model_name='analysisjob',
+            name='study',
+            field=models.ForeignKey(db_column='STUDY_ID', on_delete=django.db.models.deletion.CASCADE, related_name='analysis', to='emgapi.study'),
+        ),
+        # migrations.RunSQL(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:907]: Dev derbis

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 7]
```diff
@@ -0,0 +1,2489 @@
...
+            model_name='sample',
+            name='primary_accession',
+            field=models.CharField(db_column='PRIMARY_ACCESSION', max_length=20),
+        ),
+        migrations.AddField(
+            model_name='analysisjob',
+            name='secondary_accession',
+            field=models.CharField(db_column='SECONDARY_ACCESSION', max_length=100),
+        ),
+        # migrations.RunSQL(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:894]: Dev derbis

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 8]
```diff
@@ -0,0 +1,2489 @@
...
+            model_name='analysisjob',
+            name='secondary_accession',
+            field=models.CharField(blank=True, db_column='SECONDARY_ACCESSION', max_length=100, null=True),
+        ),
+        migrations.AlterField(
+            model_name='analysisjob',
+            name='job_operator',
+            field=models.CharField(blank=True, db_column='JOB_OPERATOR', max_length=15, null=True),
+        ),
+        # migrations.AlterField(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1079]: Dev derbis

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 9]
```diff
@@ -0,0 +1,2489 @@
...
+        ),
+        migrations.AlterModelOptions(
+            name='analysisjobdownload',
+            options={'ordering': ('pipeline', 'group_type', 'alias')},
+        ),
+        migrations.AlterModelOptions(
+            name='studydownload',
+            options={'ordering': ('pipeline', 'group_type', 'alias')},
+        ),
+        # migrations.RunSQL(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1139]: dev debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 10]
```diff
@@ -0,0 +1,2489 @@
...
+        migrations.AlterUniqueTogether(
+            name='assembly',
+            unique_together={('accession', 'wgs_accession', 'legacy_accession'), ('assembly_id', 'accession')},
+        ),
+        migrations.AddField(
+            model_name='analysisjob',
+            name='assembly',
+            field=models.ForeignKey(blank=True, db_column='ASSEMBLY_ID', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='analyses', to='emgapi.assembly'),
+        ),
+        # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1227]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 11]
```diff
@@ -0,0 +1,2489 @@
...
+                'indexes': [],
+            },
+            bases=('emgapi.sample',),
+        ),
+        migrations.AlterField(
+            model_name='analysisjob',
+            name='experiment_type',
+            field=models.ForeignKey(db_column='EXPERIMENT_TYPE_ID', on_delete=django.db.models.deletion.CASCADE, related_name='analyses', to='emgapi.experimenttype'),
+        ),
+        # migrations.AlterField(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1250]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 12]
```diff
@@ -0,0 +1,2489 @@
...
+            model_name='sample',
+            name='species',
+            field=models.CharField(blank=True, db_column='SPECIES', help_text='Species', max_length=255, null=True),
+        ),
+        migrations.AlterField(
+            model_name='sampleann',
+            name='sample',
+            field=models.ForeignKey(db_column='SAMPLE_ID', on_delete=django.db.models.deletion.CASCADE, related_name='metadata', to='emgapi.sample'),
+        ),
+        # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1425]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 13]
```diff
@@ -0,0 +1,2489 @@
...
+        #     code=rm_sample_ann_optional_keys,
+        # ),
+        migrations.AlterUniqueTogether(
+            name='sampleann',
+            unique_together=set(),
+        ),
+        # migrations.AddField(
+        #     model_name='sampleann',
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1433]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 14]
```diff
@@ -0,0 +1,2489 @@
...
+        migrations.AlterUniqueTogether(
+            name='sampleann',
+            unique_together={('sample', 'var')},
+        ),
+        migrations.AlterField(
+            model_name='studypublication',
+            name='study',
+            field=models.ForeignKey(db_column='STUDY_ID', on_delete=django.db.models.deletion.CASCADE, to='emgapi.study'),
+        ),
+        # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1459]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 15]
```diff
@@ -0,0 +1,2489 @@
...
+                ('parent_id', models.ForeignKey(blank=True, db_column='PARENT_DOWNLOAD_ID', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='parent', to='emgapi.genomedownload')),
+                ('subdir', models.ForeignKey(blank=True, db_column='SUBDIR_ID', null=True, on_delete=django.db.models.deletion.CASCADE, to='emgapi.downloadsubdir')),
+            ],
+            options={
+                'db_table': 'GENOME_DOWNLOAD',
+                'ordering': ('group_type', 'alias'),
+                'unique_together': {('realname', 'alias', 'genome')},
+            },
+        ),
+        # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1743]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 16]
```diff
@@ -0,0 +1,2489 @@
...
+            reverse_sql='ALTER TABLE SUPER_STUDY DROP INDEX super_study_title_ts_idx',
+        ),
+        RunSqlIfMySQL(
+            sql='CREATE FULLTEXT INDEX super_study_description_ts_idx ON SUPER_STUDY(description)',
+            reverse_sql='ALTER TABLE SUPER_STUDY DROP INDEX super_study_description_ts_idx',
+        ),
+        migrations.RunPython(
+            code=create_pipeline_v5,
+        ),
+        # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1823]: debris

File: emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py [hunk 17]
```diff
@@ -0,0 +1,2489 @@
...
+            name='cmseq',
+            field=models.FloatField(db_column='CMSEQ', null=True),
+        ),
+        migrations.AddField(
+            model_name='genome',
+            name='taxincons',
+            field=models.FloatField(db_column='TAXINCONS', default=0.0),
+            preserve_default=False,
+        ),
+        # # migrations.RunPython(
```
Comment (mberacochea) [inline emgapi/migrations/0001_squashed_0043_auto_20220811_1308.py:1889]: debris

File: fixtures/analysis_statuses.json [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    "ANALYSIS_STATUS_ID": 3,
+    "ANALYSIS_STATUS": "completed"
+  },
+  {
+    "ANALYSIS_STATUS_ID": 4,
+    "ANALYSIS_STATUS": "failed"
+  },
+  {
+    "ANALYSIS_STATUS_ID": 5,
+    "ANALYSIS_STATUS": "suppressed"
```
Comment (mberacochea) [inline fixtures/analysis_statuses.json:20]: Didn't we removed this with `is_suppressed`?

**Review by mberacochea — APPROVED** [review]
Great stuff!! Thank you Sandy.

File: fixtures/analysis_statuses.json (hunk 18 above)
Comment (SandyRogers) [reply fixtures/analysis_statuses.json:20]: > Didn't we removed this with `is_suppressed`?

That row is still present in EMG REL DB – possibly nothing fk's to it, but the goal here is to populate an empty DB to match that.

**Pull request merged**